File: shared/src/commonMain/kotlin/energy/ftw/crypto/Transport.kt
package energy.ftw.crypto

const val SEQ_BYTES = 8
const val TRANSPORT_OVERHEAD = SEQ_BYTES + TAG_BYTES
const val REPLAY_WINDOW = 64

class NoiseTransport(result: HandshakeResult) {
    val handshakeHash: ByteArray = result.handshakeHash
    val remoteStatic: ByteArray = result.remoteStatic
    private val send = result.send
    private val recv = result.recv
    private var highestSeq = -1L
    private var seen = 0L
    private var closed = false

    val nextSeq: Long get() = send.nonce

    fun encrypt(frame: ByteArray): ByteArray {
        assertOpen()
        val seq = writeSeq(send.nonce)
        val body = send.encryptWithAd(seq, frame)
        return concatSeq(seq, body)
    }

    fun decrypt(bytes: ByteArray): ByteArray {
        assertOpen()
        if (bytes.size < TRANSPORT_OVERHEAD) {
            throw NoiseError("transport message is ${bytes.size} bytes", "E_NOISE_MESSAGE")
        }
        val seqBytes = bytes.copyOfRange(0, SEQ_BYTES)
        val seq = readSeq(seqBytes)
        checkReplay(seq)
        recv.setNonce(seq)
        val frame = recv.decryptWithAd(seqBytes, bytes.copyOfRange(SEQ_BYTES, bytes.size))
        markSeen(seq)
        return frame
    }

    fun close() {
        closed = true
        send.destroy()
        recv.destroy()
    }

    private fun assertOpen() {
        if (closed) throw NoiseError("session is closed", "E_NOISE_CLOSED")
    }

    private fun checkReplay(seq: Long) {
        if (seq > highestSeq) return
        val behind = highestSeq - seq
        if (behind >= REPLAY_WINDOW) {
            throw NoiseError("sequence $seq is outside the replay window", "E_NOISE_REPLAY")
        }
        if (((seen ushr behind.toInt()) and 1L) == 1L) {
            throw NoiseError("sequence $seq was already accepted", "E_NOISE_REPLAY")
        }
    }

    private fun markSeen(seq: Long) {
        if (seq > highestSeq) {
            val shift = seq - highestSeq
            seen = if (shift >= REPLAY_WINDOW) 1L else ((seen shl shift.toInt()) or 1L) and WINDOW_MASK
            highestSeq = seq
        } else {
            seen = seen or (1L shl (highestSeq - seq).toInt())
        }
    }

    companion object {
        private const val WINDOW_MASK = (1L shl REPLAY_WINDOW) - 1L
    }
}

private fun writeSeq(seq: Long): ByteArray {
    val out = ByteArray(SEQ_BYTES)
    var v = seq
    for (i in SEQ_BYTES - 1 downTo 0) {
        out[i] = (v and 0xff).toByte()
        v = v ushr 8
    }
    return out
}

private fun readSeq(bytes: ByteArray): Long {
    var v = 0L
    for (i in 0 until SEQ_BYTES) v = (v shl 8) or (bytes[i].toLong() and 0xff)
    return v
}

private fun concatSeq(seq: ByteArray, body: ByteArray): ByteArray {
    val out = ByteArray(seq.size + body.size)
    seq.copyInto(out)
    body.copyInto(out, seq.size)
    return out
}

File: appleApp/FTWKit/Sources/FTWKit/Crypto/Transport.swift
import Foundation

/// The encrypted channel once the handshake is done.
///
///     offset  field       note
///     0       seq   u64BE  cleartext, authenticated as associated data
///     8       body        ChaCha20-Poly1305 over the frame, tag included
///
/// The sequence number is on the wire so a carrier that reorders or drops
/// (the LAN path, later) costs one frame instead of the session. Eight bytes
/// of monotonic counter and nothing else, so frame size stays constant.
public final class NoiseTransport {
    public static let seqBytes = 8
    public static let overhead = seqBytes + Noise.tagBytes
    /// At 1 Hz on lane 0, 64 frames is a minute of tolerance and one word.
    public static let replayWindow: UInt64 = 64

    public let handshakeHash: Bytes
    /// The box's static key as authenticated, not as claimed.
    public let remoteStatic: Bytes

    private let send: CipherState
    private let recv: CipherState
    private var highestSeq: UInt64?
    private var seen: UInt64 = 0
    private var closed = false

    public init(_ result: HandshakeResult) {
        send = result.send
        recv = result.recv
        handshakeHash = result.handshakeHash
        remoteStatic = result.remoteStatic
    }

    public var nextSeq: UInt64 { send.nonce }

    public func encrypt(_ frame: Bytes) throws -> Bytes {
        try assertOpen()
        let seq = send.nonce.bigEndianBytes
        let body = try send.encrypt(ad: seq, frame)
        return seq + body
    }

    public func decrypt(_ bytes: Bytes) throws -> Bytes {
        try assertOpen()
        guard bytes.count >= Self.overhead else {
            throw Noise.NoiseError(code: "E_NOISE_MESSAGE", message: "transport message is \(bytes.count) bytes")
        }
        let seqBytes = Array(bytes[0..<Self.seqBytes])
        let seq = seqBytes.reduce(UInt64(0)) { $0 << 8 | UInt64($1) }
        try checkReplay(seq)
        try recv.setNonce(seq)
        // The sequence number is authenticated with the frame, so nobody can
        // renumber a captured message into the window.
        let frame = try recv.decrypt(ad: seqBytes, Array(bytes[Self.seqBytes...]))
        // Only an authenticated frame moves the window; otherwise anyone who
        // can inject bytes could burn numbers and lock out the real box.
        markSeen(seq)
        return frame
    }

    /// Drop both keys. After this the session is unusable, which is the point.
    public func close() {
        closed = true
        send.destroy()
        recv.destroy()
    }

    private func assertOpen() throws {
        if closed { throw Noise.NoiseError(code: "E_NOISE_CLOSED", message: "session is closed") }
    }

    private func checkReplay(_ seq: UInt64) throws {
        guard let highest = highestSeq, seq <= highest else { return }
        let behind = highest - seq
        if behind >= Self.replayWindow {
            throw Noise.NoiseError(code: "E_NOISE_REPLAY", message: "sequence \(seq) is outside the replay window")
        }
        if (seen >> behind) & 1 == 1 {
            throw Noise.NoiseError(code: "E_NOISE_REPLAY", message: "sequence \(seq) was already accepted")
        }
    }

    private func markSeen(_ seq: UInt64) {
        guard let highest = highestSeq else {
            highestSeq = seq
            seen = 1
            return
        }
        if seq > highest {
            let shift = seq - highest
            seen = shift >= Self.replayWindow ? 1 : (seen << shift) | 1
            highestSeq = seq
        } else {
            seen |= 1 << (highest - seq)
        }
    }
}
